pkg/model/api/v1/secret/oauth.go

```diff
@@ -0,0 +1,154 @@
+package secret
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+)
+
+type PublicOAuth struct {
+	httpClient       *http.Client
+	tlsConfig        *TLSConfig
+	ClientID         string           `json:"client_id" yaml:"client_id"`
+	ClientSecret     Hidden           `json:"client_secret" yaml:"client_secret"`
+	ClientSecretFile string           `json:"client_secret_file" yaml:"client_secret_file"`
+	TokenURL         string           `json:"token_url" yaml:"token_url"`
+	Scopes           []string         `json:"scopes" yaml:"scopes"`
+	EndpointParams   url.Values       `json:"endpoint_params" yaml:"endpoint_params"`
+	AuthStyle        oauth2.AuthStyle `json:"auth_style" yaml:"auth_style"`
+}
+
+func NewPublicOAuth(oauth *OAuth, tlsConfig *TLSConfig) *PublicOAuth {
+	if oauth == nil {
+		return nil
+	}
+
+	return &PublicOAuth{
+		tlsConfig:        tlsConfig,
+		ClientID:         oauth.ClientID,
+		ClientSecret:     Hidden(oauth.ClientSecret),
+		ClientSecretFile: oauth.ClientSecretFile,
+		TokenURL:         oauth.TokenURL,
+		Scopes:           oauth.Scopes,
+		EndpointParams:   oauth.EndpointParams,
+		AuthStyle:        oauth.AuthStyle,
+	}
+}
+
+type OAuth struct {
+	token            *oauth2.Token
+	tlsConfig        *TLSConfig
+	ClientID         string           `json:"client_id" yaml:"client_id"`
+	ClientSecret     string           `json:"client_secret" yaml:"client_secret"`
+	ClientSecretFile string           `json:"client_secret_file" yaml:"client_secret_file"`
+	TokenURL         string           `json:"token_url" yaml:"token_url"`
+	Scopes           []string         `json:"scopes" yaml:"scopes"`
+	EndpointParams   url.Values       `json:"endpoint_params" yaml:"endpoint_params"`
+	AuthStyle        oauth2.AuthStyle `json:"auth_style" yaml:"auth_style"`
+}
+
+func (o *OAuth) UnmarshalJSON(data []byte) error {
+	var tmp OAuth
+	type plain OAuth
+	if err := json.Unmarshal(data, (*plain)(&tmp)); err != nil {
+		return err
+	}
+	if err := (&tmp).validate(); err != nil {
+		return err
+	}
+	*o = tmp
+	return nil
+}
+
+func (o *OAuth) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var tmp OAuth
+	type plain OAuth
+	if err := unmarshal((*plain)(&tmp)); err != nil {
+		return err
+	}
+	if err := (&tmp).validate(); err != nil {
+		return err
+	}
+	*o = tmp
+	return nil
+}
+
+// GetToken exchanges the client credentials for an access token,
+// from the OAuth 2.0 provider.
+func (o *OAuth) GetToken(ctx context.Context) (*oauth2.Token, error) {
+	httpClient, err := newHTTPClient(o.tlsConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// add out http client with tls config
+	context.WithValue(ctx, oauth2.HTTPClient, httpClient)
+
+	if o.ClientSecret == "" && o.ClientSecretFile != "" {
```

Comment:
for this part, I think it won't hurt if `oauth` has a function to get the `clientSecret`, like what we are providing in the secret.Authorization

```go
func (o *OAuth) GetClientSecret() (string, error) {
	if len(o.ClientSecretFile) > 0 {
		data, err := os.ReadFile(o.ClientSecretFile)
		if err != nil {
			return "", err
		}
		return string(data), nil
	}
	return a.ClientSecret, nil
}
```